src/coreclr/nativeaot/System.Private.TypeLoader/src/Internal/Runtime/TypeLoader/TypeLoaderEnvironment.StaticsLookup.cs

```diff
@@ -197,62 +191,64 @@ public IntPtr TryGetThreadStaticFieldData(RuntimeTypeHandle runtimeTypeHandle)
+                        return typeAsEEType->DynamicThreadStaticsIndex;
+                }
+            }
 
             // Type has no GC statics
             return IntPtr.Zero;
         }
 
-        public int TryGetThreadStaticsSizeForDynamicType(int index, out int numTlsCells)
+        public IntPtr GetThreadStaticGCDescForDynamicType(TypeManagerHandle typeManagerHandle, uint index)
         {
-            Debug.Assert((index & DynamicTypeTlsOffsetFlag) == DynamicTypeTlsOffsetFlag);
-
-            numTlsCells = _maxTlsCells;
-
             using (LockHolder.Hold(_threadStaticsLock))
             {
-                int storageSize;
-                if (_dynamicGenericsThreadStaticSizes.TryGetValue((uint)index, out storageSize))
-                    return storageSize;
+                return _dynamicGenericsThreadStaticDescs[typeManagerHandle.GetIntPtrUNSAFE()][index];
             }
-
-            Debug.Assert(false);
-            return 0;
         }
 
-        public uint GetNextThreadStaticsOffsetValue()
+        public uint GetNextThreadStaticsOffsetValue(TypeManagerHandle typeManagerHandle)
         {
-            // Highest bit of the TLS offset used as a flag to indicate that it's a special TLS offset of a dynamic type
-            var result = 0x80000000 | (uint)_maxTlsCells;
-            // Use checked arithmetics to ensure there aren't any overflows/truncations
-            _maxTlsCells = checked(_maxTlsCells + 1);
+            if (!_maxThreadLocalIndex.TryGetValue(typeManagerHandle.GetIntPtrUNSAFE(), out uint result))
+                result = (uint)RuntimeAugments.GetHighestStaticThreadStaticIndex(typeManagerHandle);
+
+            _maxThreadLocalIndex[typeManagerHandle.GetIntPtrUNSAFE()] = checked(++result);
+
             return result;
         }
 
-        public void RegisterDynamicThreadStaticsInfo(RuntimeTypeHandle runtimeTypeHandle, uint offsetValue, int storageSize)
+        public void RegisterDynamicThreadStaticsInfo(RuntimeTypeHandle runtimeTypeHandle, uint offsetValue, int storageSize, IntPtr gcDesc)
         {
             bool registered = false;
             Debug.Assert(offsetValue != 0 && storageSize > 0 && runtimeTypeHandle.IsDynamicType());
 
+            IntPtr typeManager = runtimeTypeHandle.GetTypeManager().GetIntPtrUNSAFE();
+
             _threadStaticsLock.Acquire();
             try
             {
-                // Sanity check to make sure we do not register thread statics for the same type more than once
-                uint temp;
-                Debug.Assert(!_dynamicGenericsThreadStatics.TryGetValue(runtimeTypeHandle, out temp) && storageSize > 0);
-
-                _dynamicGenericsThreadStatics.Add(runtimeTypeHandle, offsetValue);
-                _dynamicGenericsThreadStaticSizes.Add(offsetValue, storageSize);
+                if (!_dynamicGenericsThreadStaticDescs.TryGetValue(typeManager, out LowLevelDictionary<uint, IntPtr> gcDescs))
```

Comment:
We have:
- `BagElementKind.GcStaticDesc` (unused)
- `BagElementKind.ThreadStaticDesc`
- `BagElementKind.GcStaticEEType`

and the naming is inconsistent.

We should unify this and all derived names to either:

- `BagElementKind.GcStaticDesc`
- `BagElementKind.ThreadStaticDesc`

or:

- `BagElementKind.GcStaticEEType`
- `BagElementKind.ThreadStaticEEType`

I think I like the first option better (GcStaticDesc + ThreadStaticDesc). These are not fullblown EE types.